src/unix/udp.c

```diff
@@ -616,23 +639,19 @@ static int uv__setsockopt_maybe_char(uv_udp_t* handle,
   if (val < 0 || val > 255)
     return -EINVAL;
 
-  if (handle->flags & UV_HANDLE_IPV6)
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IPV6,
-                   option6,
-                   &arg,
-                   sizeof(arg));
-  else
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IP,
-                   option4,
-                   &arg,
-                   sizeof(arg));
-
-  if (r)
-    return -errno;
+/*
+ * On Solaris and derivatives such as SmartOS, the length of socket options
+ * is always sizeof(int), except for the IP_MULTICAST_LOOP and IP_MULTICAST_TTL
+ * options.
+ */
+#if defined(__sun)
+  if ((handle->flags & UV_HANDLE_IPV6) ||
```

Comment:
I don't think the check for the option is sufficient, because the size of`IP_MULTICAST_TTL` is different than the size of `IPV6_MULTICAST_HOPS`, same difference for `IP_MULTICAST_LOOP` and `IPV6_MULTICAST_LOOP`.

In other words, sometimes for the same option, the size differs, but it's always `sizeof(int)` for IPv6 options. I'm actually writing a test to make sure that this statement is true for all IPv6 options. I'll include a link to it in the comments soon.

Does that make sense?